virt_lightning/api.py

```diff
@@ -0,0 +1,442 @@
+    hv.init_storage_pool(configuration.storage_pool)
+    for domain in hv.list_domains():
+        if domain.context != context:
+            continue
+        logger.info("%s purging %s", symbols.TRASHBIN.value, domain.name)
+        hv.clean_up(domain)
+
+    if bool(distutils.util.strtobool(configuration.network_auto_clean_up)):
+        hv.network_obj.destroy()
+
+
+def distro_list(configuration, **kwargs):
+    """
+    Return a list of VM images that are available on the system.
+    """
+    conn = libvirt.open(configuration.libvirt_uri)
+    hv = vl.LibvirtHypervisor(conn)
+    hv.init_storage_pool(configuration.storage_pool)
+    return hv.distro_available()
+
+
+def storage_dir(configuration, **kwargs):
+    """
+    Return the location of the VM image storage directory.
+    """
+    conn = libvirt.open(configuration.libvirt_uri)
+    hv = vl.LibvirtHypervisor(conn)
+    hv.init_storage_pool(configuration.storage_pool)
+    return hv.get_storage_dir()
+
+
+def fetch(configuration, progress_callback=None, **kwargs):
+    """
+    Retrieve a VM image from Internet.
+    """
+    conn = libvirt.open(configuration.libvirt_uri)
+    hv = vl.LibvirtHypervisor(conn)
+    hv.init_storage_pool(configuration.storage_pool)
+    storage_dir = hv.get_storage_dir()
+
+    try:
+        r = urllib.request.urlopen(
+            "https://virt-lightning.org/images/{distro}/{distro}.qcow2".format(**kwargs)
+        )
+    except urllib.error.HTTPError as e:
+        if e.code == 404:
+            logger.error(
+                (
+                    "Distro %s not found!\n"
+                    "Visit https://virt-lightning.org/images/ "
+                    "to get an up to date list."
+                ),
+                kwargs["distro"],
+            )
+
+            sys.exit(1)
+        else:
+            logger.exception(e)
+            sys.exit(1)
+    lenght = int(r.headers["Content-Length"])
+    chunk_size = MB * 1
+    target_file = pathlib.PosixPath(
+        "{storage_dir}/upstream/{distro}.qcow2".format(
+            storage_dir=storage_dir, **kwargs
+        )
+    )
+    temp_file = target_file.with_suffix(".temp")
+    if target_file.exists():
+        logger.info(
+            "File already exists: {target_file}".format(target_file=target_file)
+        )
+        return
+    with temp_file.open("wb") as fd:
+        while fd.tell() < lenght:
+            chunk = r.read(chunk_size)
+            fd.write(chunk)
+            if progress_callback:
+                progress_callback(fd.tell(), lenght)
+    temp_file.rename(target_file)
+    try:
+        r = urllib.request.urlopen(
+            "https://virt-lightning.org/images/{distro}/{distro}.yaml".format(**kwargs)
+        )
+    except urllib.error.HTTPError as e:
+        if e.code == 404:
+            pass
+    with target_file.with_suffix(".yaml").open("wb") as fd:
+        fd.write(r.read())
+    logger.info("Image {distro} is ready!".format(**kwargs))
```

Comment:
It's usually a good practice to add `__all__` to public modules, so then `from virt_lightning.api import *` won;t bring up unwanted functions :  https://docs.python.org/3/tutorial/modules.html#importing-from-a-package